### PR TITLE
Read the language from the element's class

### PR DESCRIPTION
**Pull request by icio**
Hey @johannestroeger! We use your gulp-highlight on HTML generated by Jekyll from markdown files, and often in our original markdown we include a Github code fence with an explicit language. For example:

    ```bash
    grep -Ro ident src
    ```

Since we using the Kramdown renderer (with its own highlighting disabled) this generates HTML of the form:

```html
<pre><code class="language-bash">grep -Ro ident src</code></pre>
```

In this pull request I change the call to highlight.js such that the first class on the element equal to a language of highlight.js (or prefixed by the new `languageClassPrefix` option and suffixed by a language of highlight.js) is explicitly used in rendering. If none is found then we fall back to `highAuto` as was initially the behaviour.

This means that in the example above (where I would set `languageClassPrefix: "language-"`) we end up setting the HTML of the `code` element to `hljs.highlight('bash', elem.text()).value` where previously it was `hljs.highlightAuto(elem.html()).value`.

Note that this might be a backwards compatible change if users are highlighting on elements such as `<code class="json">` but still expect `highlightAuto` to be used. Not sure if that's a deal breaker - maybe just warrants a major version bump?

Hope this is useful.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,10 +19,17 @@ module.exports = function (options) {
...
-        elem.addClass('hljs');
+      if (elem.hasClass(options.ignoreClass)) {
+        return;
       }
+
+      var lang = detectLang(elem, options.languageClassPrefix);
+      elem.html(
+        lang
+          ? hljs.highlight(lang, elem.text()).value
+          : hljs.highlightAuto(elem.text()).value
```
Comment (icio) [inline index.js]: I'm not sure why I had to change from `elem.html()` to `elem.text()`, but I was otherwise finding that my output was double-escaped.

**Comment by icio** [comment]
Just noticed that the tests are failing - with the reference to the same encoding confusion I was asking about. Will look into it...

**Comment by jlmakes** [comment]
We definitely need support for explicit language types. Automatic language detection is good, but can fail in weird places. Look at these JavaScript comments for example:

screen shot 2017-05-28 at 8 25 27 pm

There’s no way to force Javascript syntax highlighting!

I think a languageSubset option is nice, but I use HTML, CSS and JS syntax highlighting in my website's documentation; I can’t just hard code one in the options.

We need to have support for detecting the `.lang-js` or `.javascript` classes on the `<code>` tags as described in the highlight.js documentation.

**Comment by jlmakes** [comment]
@icio You're using `String.prototype.startsWith` which is not available before Node v4.

The CI is running Node v0.10.48, which is safe to consider outdated. So, we should probably update the `.travis.yml` and bump the library to `v3.0.0`. That should at least help fix 3 of the 4 failings tests.

I'm not sure what to suggest about the first failing test and the escaped character funny business...

File: index.js [hunk 2]
```diff
@@ -47,3 +54,20 @@ module.exports = function (options) {
     cb();
   });
 };
+
+function detectLang(elem, langClassPrefix) {
+  if (!langClassPrefix) langClassPrefix = '';
+
+  var classes = (elem.attr('class') || '').split(' ');
+  for (var c = 0; c < classes.length; c++) {
```
Comment (jlmakes) [inline index.js]: hehe, `c++`

**Pull request closed without merging**